src/fs_report.rs

```diff
@@ -0,0 +1,86 @@
+// Copyright 2022 Adobe. All rights reserved.
+// This file is licensed to you under the Apache License,
+// Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
+// or the MIT license (http://opensource.org/licenses/MIT),
+// at your option.
+// Unless required by applicable law or agreed to in writing,
+// this software is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR REPRESENTATIONS OF ANY KIND, either express or
+// implied. See the LICENSE-MIT and LICENSE-APACHE files for the
+// specific language governing permissions and limitations under
+// each license.
+
+use anyhow::Result;
+use c2pa::{Ingredient, Manifest, ManifestStore, ManifestStoreReport};
+use std::fs::{create_dir_all, File};
+use std::io;
+use std::io::Write;
+use std::path::Path;
+
+const THUMBNAIL_CLAIM_NAME: &str = "thumbnail_claim";
+
+// Writes the provided `manifest` thumbnail, if present, to the `destination` path.
+fn write_manifest_thumbnail(manifest: &Manifest, destination: &Path) -> Result<()> {
+    manifest.thumbnail().map(|(format, bytes)| {
+        let name = match format {
+            "image/jpg" | "image/jpeg" => format!("{}.{}", THUMBNAIL_CLAIM_NAME, "jpeg"),
+            "image/png" => format!("{}.{}", THUMBNAIL_CLAIM_NAME, "png"),
+            _ => THUMBNAIL_CLAIM_NAME.to_owned(),
+        };
+        File::create(destination.join(name))?.write_all(bytes)
+    });
+    Ok(())
+}
+
+// Writes each of the provided `ingredient`'s thumbnails, if present, to the `destination` path.
+fn write_ingredient_thumbnails(ingredients: &[Ingredient], destination: &Path) -> Result<()> {
+    ingredients
+        .iter()
+        .filter_map(|ingredient| {
+            ingredient
+                .thumbnail()
+                .map(|(_, bytes)| (ingredient.title(), bytes))
+        })
+        .map(|(title, bytes)| File::create(destination.join(title))?.write_all(bytes))
```

Comment:
The thumbnail is not always in the same format as the ingredient, so we can't just directly use the ingredient title here. We would need to replace the title extension with the correct extension type from the thumbnail content type